judge/views/problem.py

```diff
@@ -201,6 +201,95 @@ def get_context_data(self, **kwargs):
                                           context['description'], 'problem')
         context['meta_description'] = self.object.summary or metadata[0]
         context['og_image'] = self.object.og_image or metadata[1]
+
+        context['can_vote'] = self.object.can_vote(user)
+        if context['can_vote']:
+            try:
+                context['vote'] = ProblemPointsVote.objects.get(voter=user.profile, problem=self.object)
+            except ObjectDoesNotExist:
+                context['vote'] = None
+        else:
+            context['vote'] = None
+
+        return context
+
+
+class ProblemVote(ProblemMixin, DetailView):
+    template_name = 'problem/vote-ajax.html'
+
+    def get_context_data(self, **kwargs):
+        problem = self.get_object()
+        if not problem.can_vote(self.request.user):
+            raise Http404()
+
+        context = super(ProblemVote, self).get_context_data(**kwargs)
+
+        context['problem'] = problem
```

Comment:
This feels awkward. `context['problem']` should be set by `DetailView` if you add `context_object_name = 'problem'`. I think we should do that instead. As for the check on line 222, I think what we usually do is we have it in `get_object()` instead (see `ProblemMixin`).

Reply:
- Removed `context['problem']`.
- Both `ProblemMixin.get_object` and `can_vote` are needed. `ProblemMixin.get_object` checks if the user can access the problem. `can_vote` does additional checks to see if the user can vote.